### PR TITLE
Support for TIME(p) and TIMESTAMP(p) to SQLAlchemy

### PR DESCRIPTION
**Pull request by hovaesco**
First point from https://github.com/trinodb/trino-python-client/issues/107

### CODE REVIEW

File: trino/sqlalchemy/compiler.py [hunk 1]
```diff
@@ -147,6 +147,39 @@ def visit_BLOB(self, type_, **kw):
     def visit_DATETIME(self, type_, **kw):
         return self.visit_TIMESTAMP(type_, **kw)
 
+    def visit_INTERVAL(self, type_, **kw):
```
Comment (hovaesco) [inline trino/sqlalchemy/compiler.py:150]: I'm not really sure whether it's a correct approach. Interval in Trino has a different value and data type. 
For instance:
- data type: `INTERVAL YEAR TO MONTH`
- value: `INTERVAL '3' MONTH`

File: tests/unit/sqlalchemy/test_datatype_parse.py [hunk 2]
```diff
@@ -173,12 +176,18 @@ def test_parse_row(type_str: str, sql_type: ARRAY, assert_sqltype):
...
-    # TODO: support parametric timestamps (https://github.com/trinodb/trino-python-client/issues/107)
     "date": DATE(),
     "time": TIME(),
+    "time(3)": TIME(3, timezone=False),
+    "time(6)": TIME(6),
+    "time(12) with time zone": TIME(12, timezone=True),
     "time with time zone": TIME(timezone=True),
-    "timestamp": TIMESTAMP(),
+    "timestamp(3)": TIMESTAMP(3, timezone=False),
+    "timestamp(6)": TIMESTAMP(6),
```
Comment (andrewdibiasio6) [inline tests/unit/sqlalchemy/test_datatype_parse.py:186]: are you missing timezone here?

File: trino/sqlalchemy/compiler.py (hunk 1 above)
Comment (andrewdibiasio6) [reply trino/sqlalchemy/compiler.py:150]: Is this func supposed to return the data type or value? It seems like it returns a mix of both.

File: trino/sqlalchemy/compiler.py [hunk 3]
```diff
@@ -147,6 +147,26 @@ def visit_BLOB(self, type_, **kw):
...
 
+    def visit_TIMESTAMP(self, type_, **kw):
+        return "TIMESTAMP%s%s" % (
+            "(%d)" % type_.precision
+            if getattr(type_, "precision", None) is not None
+            else "",
+            " WITH TIME ZONE"
+            if getattr(type_, "timezone", False)
+            else ""
+        )
```
Comment (dungdm93) [inline trino/sqlalchemy/compiler.py:158]: I prefer build string instead

```python
    def visit_TIMESTAMP(self, type_, **kw):
        datatype = "TIMESTAMP"
        precision = getattr(type_, "precision", None)
        if precision:
            datatype += f"({precision})"
        if getattr(type_, "timezone", False):
            datatype += " WITH TIME ZONE"
        
        return datatype
```

**Review by dungdm93 — CHANGES_REQUESTED** [review]

File: trino/sqlalchemy/datatype.py [hunk 4]
```diff
@@ -193,7 +211,9 @@ def parse_sqltype(type_str: str) -> TypeEngine:
...
     type_args = [int(o.strip()) for o in type_opts.split(",")] if type_opts else []
     if type_name in ("time", "timestamp"):
-        type_kwargs = dict(timezone=type_str.endswith("with time zone"))
-        # TODO: support parametric timestamps (https://github.com/trinodb/trino-python-client/issues/107)
-        return type_class(**type_kwargs)
+        if type_str.endswith("with time zone"):
+            type_kwargs: Dict[str, Any] = dict(timezone=True)
+            if type_opts is not None:
+                type_kwargs["precision"] = int(type_opts)
+            return type_class(**type_kwargs)
```
Comment (dungdm93) [inline trino/sqlalchemy/datatype.py:218]: I expect this `if` block handle all cases of `time` & `timestamp` datatype:
```python
    if type_name in ("time", "timestamp"):
        type_kwargs: Dict[str, Any] = dict()
        if type_str.endswith("with time zone"):
            type_kwargs["timezone"] = True
        if type_opts is not None:
            type_kwargs["precision"] = int(type_opts)
        return type_class(**type_kwargs)
```

**Review by dungdm93 — CHANGES_REQUESTED** [review]

File: trino/sqlalchemy/compiler.py (hunk 3 above)
Comment (hovaesco) [reply trino/sqlalchemy/compiler.py]: done for both TIME and TIMESTAMP

File: trino/sqlalchemy/datatype.py (hunk 4 above)
Comment (hovaesco) [reply trino/sqlalchemy/datatype.py]: done

**Review by dungdm93 — APPROVED** [review]

**Review by ebyhr — COMMENTED** [review]
Could you please provide simple test script? As you already know, we don't have integration tests for sqlalchemy. I hesitate to merge with confidence.

**Comment by ebyhr** [comment]
@hovaesco Thanks for providing the code. Could you test with `precision=0`? It creates `time(3)` now, but it should be `time(0)`.

File: trino/sqlalchemy/compiler.py [hunk 5]
```diff
@@ -147,6 +147,29 @@ def visit_BLOB(self, type_, **kw):
     def visit_DATETIME(self, type_, **kw):
         return self.visit_TIMESTAMP(type_, **kw)
 
+    def visit_TIMESTAMP(self, type_, **kw):
+        datatype = "TIMESTAMP"
+        precision = getattr(type_, "precision", None)
+        if precision not in range(0, 12):
```
Comment (dungdm93) [inline trino/sqlalchemy/compiler.py:153]: `range(0,12)` mean `[0, 12)`
Trino is support precision up to `12` (picoseconds), [according the docs](https://trino.io/docs/current/language/types.html#time-p).

Comment (hovaesco) [reply trino/sqlalchemy/compiler.py]: ohh.. thanks for catching it

**Comment by andrewdibiasio6** [comment]
LGTM pending ebyhr approval.

**Review by andrewdibiasio6 — APPROVED** [review]

File: trino/sqlalchemy/compiler.py [hunk 6]
```diff
@@ -147,6 +147,29 @@ def visit_BLOB(self, type_, **kw):
     def visit_DATETIME(self, type_, **kw):
         return self.visit_TIMESTAMP(type_, **kw)
 
+    def visit_TIMESTAMP(self, type_, **kw):
+        datatype = "TIMESTAMP"
+        precision = getattr(type_, "precision", None)
+        if not 0 <= precision <= 12:
+            raise ValueError(f"invalid precision={precision}, it must be from range 0-12")
+        if precision is not None:
```
[the remark this answers is not in the document]
Comment (dungdm93) [reply trino/sqlalchemy/compiler.py:155]: NPE. LOL

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Merge, thanks!

**Pull request merged**